### PR TITLE
feat: Add support for dual-stack IP addresses - IPv4 + IPv6.

### LINKED ISSUE #112: Add support for dual-stack IP addresses (both ipv4 and v6)
Looking at the terraform code in main.tf:
https://github.com/terraform-google-modules/terraform-google-lb-http/blob/master/main.tf
there is provision to specify only 1 version of IP addresses via the `ip_version` variable (defaulting to IPv4) in the `google_compute_global_address` resource (https://github.com/terraform-google-modules/terraform-google-lb-http/blob/master/main.tf#L42)

Even when the user creates external static IP addresses, there is provision to supply only one such value for the `google_compute_global_forwarding_rule` resources using `local.address`.

In the transitory period of IP versions, it's important to have both IPv4 **and** IPv6 addresses assigned to the load-balancer.

One option to make it configurable would be to accept the variable `ip_version` as a list `ip_versions` where the user can specify what versions of IP addresses they want to deploy - something like:
```
    ip_versions = ["IPv4", "IPv6"]
```
and create the corresponding `google_compute_global_address` resources based on that.
It can default to `["IPv4"]` to retain backwards compatibility.

### PR DESCRIPTION
**Pull request by kunal-g**
We need to create multiple forwarding rules -
one for each (ip version address, port) combination - all pointing to
the same proxies:

* ipv4 addr + http/80 - pointing to the **http** proxy
* ipv4 addr + https/443 - pointing to the **https** proxy
* ipv6 addr + http/80 - pointing to the **http** proxy
* ipv6 addr + https/443 - pointing to the **https** proxy

Fixes #112.

Signed-off-by: Kunal Gangakhedkar <kunalg@hotstar.com>

### CODE REVIEW

**Comment by kunal-g** [comment]
@morgante can you please help with what's failing in the checks?
I cannot see the lint failures - no permissions.

**Comment by morgante** [comment]
@kunal-g Looks like the issue is based on #117

**Comment by kunal-g** [comment]
> @kunal-g Looks like the issue is based on #117

Cool, so no action-items for me?

**Comment by kunal-g** [comment]
> @kunal-g Looks like the issue is based on #117

There is one more fix needed - please check comment https://github.com/terraform-google-modules/terraform-google-lb-http/pull/99#issuecomment-681809565

**Comment by kunal-g** [comment]
@morgante any updates on this PR?

**Comment by morgante** [comment]
@kunal-g Please get the tests passing.

Linter issues:
```
Checking submodule's files generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' /workspace/README.md /tmp/tmp.RigFUGhxSB/workspace/README.md
74c74
<       iap_config = {
---
>       iap_config {
Error: submodule's files generation has not been run, please run the
'make build' command and commit changes
/usr/local/bin/task_helper_functions.sh: line 30: DELETE_AT_EXIT: readonly variable
Checking for documentation generation
Checking for trailing whitespace
Checking for missing newline at end of file
Running shellcheck
Checking file headers
Running flake8
Running terraform fmt
Running terraform validate
terraform_validate . 
Success! The configuration is valid.

terraform_validate ./examples/https-gke 
Success! The configuration is valid.

terraform_validate ./examples/https-gke/gke-node-port 
Success! The configuration is valid.

terraform_validate ./examples/https-redirect 

Error: Invalid value for module argument

  on main.tf line 104, in module "gce-lb-http":
 104:   backends = {
 105:     default = {
 106:       description                     = null
 107:       protocol                        = "HTTP"
 108:       port                            = 80
 109:       port_name                       = "http"
 110:       timeout_sec                     = 10
 111:       connection_draining_timeout_sec = null
 112:       enable_cdn                      = false
 113:       session_affinity                = null
 114:       affinity_cookie_ttl_sec         = null
 116:       health_check = {
 117:         check_interval_sec  = null
 118:         timeout_sec         = null
 119:         healthy_threshold   = null
 120:         unhealthy_threshold = null
 121:         request_path        = "/"
 122:         port                = 80
 123:         host                = null
 124:         logging             = null
 125:       }
 127:       log_config = {
 128:         enable      = false
 129:         sample_rate = null
 130:       }
 132:       groups = [
 133:         {
 134:           group                        = module.mig.instance_group
 135:           balancing_mode               = null
 136:           capacity_scaler              = null
 137:           description                  = null
 138:           max_connections              = null
 139:           max_connections_per_instance = null
 140:           max_connections_per_endpoint = null
 141:           max_rate                     = null
 142:           max_rate_per_instance        = null
 143:           max_rate_per_endpoint        = null
 144:           max_utilization              = null
 145:         }
 146:       ]
 147:     }
 148:   }

The given value is not suitable for child module variable "backends" defined
at ../../variables.tf:75,1-20: element "default": attribute "iap_config" is
required.

terraform_validate ./examples/mig-nat-http-lb 
Success! The configuration is valid.

terraform_validate ./examples/multi-backend-multi-mig-bucket-https-lb 
Success! The configuration is valid.

terraform_validate ./examples/multi-mig-http-lb 
Success! The configuration is valid.

terraform_validate ./examples/multiple-certs 
Success! The configuration is valid.

terraform_validate ./examples/shared-vpc 
Success! The configuration is valid.

terraform_validate ./modules/dynamic_backends 
Success! The configuration is valid.

terraform_validate ./test/fixtures/https-redirect 

Error: Invalid value for module argument

  on ../../../examples/https-redirect/main.tf line 104, in module "gce-lb-http":
 104:   backends = {
 105:     default = {
 106:       description                     = null
 107:       protocol                        = "HTTP"
 108:       port                            = 80
 109:       port_name                       = "http"
 110:       timeout_sec                     = 10
 111:       connection_draining_timeout_sec = null
 112:       enable_cdn                      = false
 113:       session_affinity                = null
 114:       affinity_cookie_ttl_sec         = null
 116:       health_check = {
 117:         check_interval_sec  = null
 118:         timeout_sec         = null
 119:         healthy_threshold   = null
 120:         unhealthy_threshold = null
 121:         request_path        = "/"
 122:         port                = 80
 123:         host                = null
 124:         logging             = null
 125:       }
 127:       log_config = {
 128:         enable      = false
 129:         sample_rate = null
 130:       }
 132:       groups = [
 133:         {
 134:           group                        = module.mig.instance_group
 135:           balancing_mode               = null
 136:           capacity_scaler              = null
 137:           description                  = null
 138:           max_connections              = null
 139:           max_connections_per_instance = null
 140:           max_connections_per_endpoint = null
 141:           max_rate                     = null
 142:           max_rate_per_instance        = null
 143:           max_rate_per_endpoint        = null
 144:           max_utilization              = null
 145:         }
 146:       ]
 147:     }
 148:   }

The given value is not suitable for child module variable "backends" defined
at ../../../variables.tf:75,1-20: element "default": attribute "iap_config" is
required.

terraform_validate ./test/fixtures/mig_nat 
Success! The configuration is valid.

terraform_validate ./test/fixtures/multi_certs 
Success! The configuration is valid.

terraform_validate ./test/fixtures/multi_mig 
Success! The configuration is valid.

terraform_validate ./test/setup 

Warning: Provider source not supported in Terraform v0.12

  on versions.tf line 20, in terraform:
  20:     google = {
  21:       source  = "hashicorp/google"
  22:       version = ">= 3.32, <4.0.0"
  23:     }

A source was declared for provider google. Terraform v0.12 does not support
the provider source attribute. It will be ignored.

(and one more similar warning elsewhere)
```

**Comment by kunal-g** [comment]
@morgante:
Got some time to look at this after a long time.

Apparently, the missing `iap_config` fix in `examples/https-redirect/main.tf` was already committed to `master` (commit: 0136c65dfceafcd468bfb907551cbebb016204d9)

Now, all the checks are passing :bowtie: - without any additional work on my part 😜 🤘 

Please consider this for merging to `master` and release.

File: autogen/variables.tf.tmpl
Comment (morgante) [inline autogen/variables.tf.tmpl:31]: For backwards-compatibility:

```suggestion
  default     = false
```

File: autogen/main.tf.tmpl
Comment (morgante) [inline autogen/main.tf.tmpl:54]: We probably need to make all the ipv6 resources conditional so they are only created if:
1. ipv6_address is provided
2. `create_ipv6_address` is true.

File: autogen/variables.tf.tmpl
Comment (morgante) [inline autogen/variables.tf.tmpl:42]:
```suggestion
  description = "Existing IPv6 address to use (the actual IP address value)"
```

File: autogen/variables.tf.tmpl
Comment (morgante) [inline autogen/variables.tf.tmpl:36]:
```suggestion
  description = "Existing IPv4 address (the actual IP address value)"
```

**Review by morgante — CHANGES_REQUESTED** [review]
Thanks for following up.

File: autogen/variables.tf.tmpl
Comment (kunal-g) [reply autogen/variables.tf.tmpl]: Done - fixed.

File: autogen/main.tf.tmpl
Comment (kunal-g) [reply autogen/main.tf.tmpl:56]: Makes sense - fixed.
* Added a new local boolean var `create_ipv6_resources` that is evaluated to true with
`((local.ipv6_address != nul)) && (length(local.ipv6_address)) > 0)`

`create_ipv6_resources` is then used in the fwd-rule resources to set the count to 0 or 1.

File: autogen/variables.tf.tmpl
Comment (kunal-g) [reply autogen/variables.tf.tmpl]: Updated docs as per suggestions.

File: autogen/variables.tf.tmpl
Comment (kunal-g) [reply autogen/variables.tf.tmpl]: Updated docs as per suggestions.

**Comment by kunal-g** [comment]
The integration tests failed for some reason - can you please let me know the issue?
I don't have permissions to check the cloud-build logs. 😞

**Comment by morgante** [comment]
@kunal-g Here's the error:

```
       Error: Invalid function argument
       
         on ../../../main.tf line 25, in locals:
         25:   create_ipv6_resources = ((local.ipv6_address != null) && (length(local.ipv6_address) > 0))
           |----------------
           | local.ipv6_address is null
       
       Invalid value for "value" parameter: argument must not be null.
```

**Comment by kunal-g** [comment]
> @kunal-g Here's the error:
> 
> ```
>        Error: Invalid function argument
>        
>          on ../../../main.tf line 25, in locals:
>          25:   create_ipv6_resources = ((local.ipv6_address != null) && (length(local.ipv6_address) > 0))
>            |----------------
>            | local.ipv6_address is null
>        
>        Invalid value for "value" parameter: argument must not be null.
> ```

Thanks. Let me figure out a way to fix this.

My expectation was that the `null` check in the combined conditional should have prevented this - must be something I missed in how terraform handles these conditionals.

File: autogen/main.tf.tmpl
Comment (morgante) [reply autogen/main.tf.tmpl:56]: Note that there are potential issues with dynamic dependencies if we do it this way (if users try to pass in a reference to their own resource). We might want to simply expose `use_ipv6` and `use_ipv4` variables.

Comment (kunal-g) [reply autogen/main.tf.tmpl:56]: Well, when I tried running it last on GCP, the reference (self-link) to the newly created IPv4/IPv6 address didn't work - it required **explicit** IP address value.

This is also reflected in the corresponding docs to the parameters that you're expected to pass the actual IP address value.

Comment (kunal-g) [reply autogen/main.tf.tmpl:56]: Maybe, we can think of more checks on the value? (for example, regex based string check)

Comment (morgante) [reply autogen/main.tf.tmpl:56]: Right, but even referencing the value is potentially an issue (ex. `ipv6_address = google_compute_global_address.mine.address`) as it will raise the error of count referencing a dynamic value.

Comment (kunal-g) [reply autogen/main.tf.tmpl:56]: Couple of use-cases I can think of where this configurability **may be** important:
1. bring-your-own-address - google supports it.
2. security considerations - where you might want to add firewall rules on-prem based on source/destination IP addresses - which might be difficult to do if this IP address keeps changing. So, a separately statically allocated elastic IP that's reviewed and approved by the security team helps here.

Let's fix this if we get any reports from users?

Comment (morgante) [reply autogen/main.tf.tmpl:56]: Right, I don't disagree with the use case. I'm just concerned that having a conditional based on the value is likely to raise errors and will require a backwards-incompatible release.

Comment (kunal-g) [reply autogen/main.tf.tmpl:56]: One way to look at it could be to treat IPv6 as equally essential as IPv4 - IPv4 address space is almost exhausted anyways and a lot of the mobile networks are running on IPv6 only - with IPv4 access provided via carrier-grade NATs.

If we take this approach, we can remove the conditional and force the IPv6 resources to be created/destroyed as first-class citizen instead of optional ones.

Comment (morgante) [reply autogen/main.tf.tmpl:56]: IPv6 isn't free so I don't want to force everyone to use it.

Comment (kunal-g) [reply autogen/main.tf.tmpl:56]: We already have a lot of variables to confuse the user:
1. ipv6_address - for custom address
2. create_ipv6_address - for creating new IPv6 address, if user doesn't have one already.
3. enable_ipv6 - the new one you proposed.

There is some interplay between these variables - and users are expected to keep track of them correctly.

I would prefer to actually cut down the variables controlling the same feature.

How about if we use:
1. `ipv6_address = "disable"` (default) to indicate no ipv6 configs to be deployed.
2. `ipv6_address = "new"` to indicate **creation** of the address
3. any other value as pre-created IPv6 address?

This could simplify the config the user is expected to fiddle with and we can then create the required local TF state variables based on that?

just a thought.

Comment (morgante) [reply autogen/main.tf.tmpl:56]: Sure, that makes sense to me.

Comment (kunal-g) [reply autogen/main.tf.tmpl:56]: Alright, let me rework the TF code with reduced no. of variables.

Comment (kunal-g) [reply autogen/main.tf.tmpl:70]: @morgante I have refactored some of the variables and conditionals in commit 4f2fb4f593e9495ddfa53b9aeea8314dbf9b904d.

Please have a look.

Essentially, I have kept only 2 user-facing variables - as per our discussion above:

1. boolean flag `enable_ipv6` to control all of the IPv6 resources
2. string var `ipv6_address` with a sentinel value `"new"` to indicate allocation of new IPv6 EIP.

Comment (morgante) [reply autogen/main.tf.tmpl:70]: Thanks. The major downside I see is that now we're referencing `ipv6_address` in a `count` conditional which means it *cannot* be a dynamic value (ex. `google_compute_global_address.mine.address`) which actually seems like a meaningful problem as I'd imagine it's a common scenario to manage the original IP address creation in Terraform.

The alternative would be to control creation with a separate `create_ipv6` variable.

That being said, i agree your interface is simpler. What do others think? @naseemkullah @bharathkkb ?

Comment (naseemkullah) [reply autogen/main.tf.tmpl:70]: I'm afraid I don't follow, what's the tl;dr as to why we cannot have `create_ipv6_address` and `ipv6_address`  to either create or use an existing ipv6 address? they would be mutually exclusive, and if both were left unset, ipv6 would be disabled.

Comment (morgante) [reply autogen/main.tf.tmpl:70]: The issue is if we want to accept `ipv6_address` as an input variable we cannot make anything conditional on it.

Comment (naseemkullah) [reply autogen/main.tf.tmpl:70]: I don't fully get it tbh, but it seems there is no working around the unfortunate third value of `enable_ipv6` if we want to be able to dynamically reference an existing `ipv6_address`, if that's true I am in favour of it since being bale to dynamically reference is worth having to deal with the confusing vars.

Comment (kunal-g) [reply autogen/main.tf.tmpl:70]: as it is, other than the **validation** of appropriate ipv6 address, I don't think we need to make the resources conditional on the variable `ipv6_address`. Do you think if we just go with `enable_ipv6` for all the conditionals would make it better/easier?

terraform doesn't seem to have a built-in ability similar to an `assert` or to raise an error and abort if validations fail - there are workarounds discussed here https://github.com/hashicorp/terraform/issues/15469.

@morgante @naseemkullah please let me know if you guys feel comfortable with only using `enable_ipv6` for conditionals and think about validating the ipv6 values later.
I'll adjust the tf code accordingly.

Comment (naseemkullah) [reply autogen/main.tf.tmpl:70]: If tf apply with an invalid ipv6 will fail on GCP's side, that is good enough IMO

Comment (kunal-g) [reply autogen/main.tf.tmpl:67]: I have reworked the `count` conditionals to not depend on the `ipv6_address` in commit 8ed7b020f50612c3bcfd6718dcc4b16a122b6e22.

@morgante @naseemkullah please have a look and let me know if this works for you.

Comment (morgante) [reply autogen/main.tf.tmpl:67]: Unfortunately the `google_compute_global_address` resource itself (line 86) is still conditional on the `ipv6_address` variable. It might be good to verify this yourself by updating one of the examples/tests to create an address outside the module then attempt to pass it into the module.

Comment (kunal-g) [reply autogen/main.tf.tmpl:67]: @morgante I have added some examples here copied from the `https-redirect` example from this repo:
https://github.com/kunal-g/google-lb-http-ipv6-examples

TL;DR:
both situations seem to be working as expected:
* predefined IPv6 address (created using `gcloud compute address create` **outside** of this module) 
* newly created IPv6 address (using the `google_compute_global_address` resource reference **within** this module).

Please let me know if there are more changes needed. Happy to help.

Comment (morgante) [reply autogen/main.tf.tmpl:67]: The situation I'm concerned about would be using the `google_compute_global_address` resource *outside* the module. Example:

```
resource "google_compute_global_address" "main" {
  project    = var.project
  name       = "my-outside-address"
  ip_version = "IPV6"
}

module "gce-lb-http" {
  source               = "github.com/kunal-g/terraform-google-lb-http?ref=issue-112-dual-stack"
  name                 = "ci-https-redirect"
  ipv6_address    = google_compute_global_address.main.address
  ...
}
```

Comment (kunal-g) [reply autogen/main.tf.tmpl:67]: Yup, you are right on target - tried what you suggested above and got this error:
```
➜  https-redirect git:(main) ✗ terraform plan -var-file=test.tfvars
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.template_file.group-startup-script: Refreshing state...
module.mig.data.google_compute_zones.available: Refreshing state...
google_compute_network.default: Refreshing state... [id=projects/frontier-268910/global/networks/default]
module.mig_template.data.google_compute_image.image_family: Refreshing state...
module.mig_template.data.google_compute_image.image: Refreshing state...
google_compute_subnetwork.default: Refreshing state... [id=projects/frontier-268910/regions/asia-south1/subnetworks/default]

Error: Invalid count argument

  on .terraform/modules/gce-lb-http/main.tf line 82, in resource "google_compute_global_address" "default_ipv6":
  82:   count      = (local.create_ipv6_address) ? 1 : 0

The "count" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the count depends on.
```

Let me figure a way out. Thanks for highlighting the case.

Comment (kunal-g) [reply autogen/main.tf.tmpl:67]: But, wouldn't the same problem occur for the `ipv4` address resources?
When I wrote the ipv6 resources, I modelled them exactly as per the current ipv4 resources.

Comment (kunal-g) [reply autogen/main.tf.tmpl:67]: Never mind, found out the reason.
So, I can only see one possible solution: go back to the 3-variable version for ipv6 addr:
1. `var.enable_ipv6` as a global switch
2. `var.create_ipv6_addr`- to create the *new* address in the module
3. `var.ipv6_address` - for own predefined address.

Does this look ok? If yes, I'll push the required changes.

Comment (morgante) [reply autogen/main.tf.tmpl:67]: Yeah I think that's the only option.

Comment (kunal-g) [reply autogen/main.tf.tmpl:60]: I have pushed the changes with 3 variables.
Please have a look and let me know if it looks OK or needs further changes.

**Comment by kunal-g** [comment]
@morgante can you please share the logs for integration test failures?

**Comment by morgante** [comment]
@kunal-g Errors look to be related to the centos image version:

```
       
       Error: error retrieving image information: googleapi: Error 404: The resource 'projects/centos-cloud/global/images/family/centos-6' was not found, notFound
       
         on .terraform/modules/example.mig2_template/modules/instance_template/main.tf line 25, in data "google_compute_image" "image_family":
         25: data "google_compute_image" "image_family" {
       
       
       
       Error: error retrieving image information: googleapi: Error 404: The resource 'projects/centos-cloud/global/images/family/centos-6' was not found, notFound
       
         on .terraform/modules/example.mig3_template/modules/instance_template/main.tf line 25, in data "google_compute_image" "image_family":
         25: data "google_compute_image" "image_family" {
       
       
```

**Comment by kunal-g** [comment]
Yeah, even I had to change the image in my examples to `centos-8`:
https://github.com/kunal-g/google-lb-http-ipv6-examples/blob/main/precreated_ipv6/https-redirect/main.tf#L90
It seems the `centos-6` image has be removed from GCP images repo.

Unfortunately, I don't know how to solve this issue. ☹️ 
Let me know if there are any action items for me.

> @kunal-g Errors look to be related to the centos image version:
> 
> ```
>        
>        Error: error retrieving image information: googleapi: Error 404: The resource 'projects/centos-cloud/global/images/family/centos-6' was not found, notFound
>        
>          on .terraform/modules/example.mig2_template/modules/instance_template/main.tf line 25, in data "google_compute_image" "image_family":
>          25: data "google_compute_image" "image_family" {
>        
>        
>        
>        Error: error retrieving image information: googleapi: Error 404: The resource 'projects/centos-cloud/global/images/family/centos-6' was not found, notFound
>        
>          on .terraform/modules/example.mig3_template/modules/instance_template/main.tf line 25, in data "google_compute_image" "image_family":
>          25: data "google_compute_image" "image_family" {
>        
>        
> ```

**Comment by morgante** [comment]
@kunal-g Could you update the `examples/` to use the latest MIG module? ex. https://github.com/terraform-google-modules/terraform-google-lb-http/blob/master/examples/multi-mig-http-lb/mig.tf#L54

**Comment by kunal-g** [comment]
> @kunal-g Could you update the `examples/` to use the latest MIG module? ex. https://github.com/terraform-google-modules/terraform-google-lb-http/blob/master/examples/multi-mig-http-lb/mig.tf#L54

Sure, but as per the docs of [latest version](https://github.com/terraform-google-modules/terraform-google-vm/releases/tag/v6.0.0), this is a breaking change and requires terraform 0.13. The earlier versions don't seem to carry the default image family update.

Do we need to maintain backwards compatibility with terraform 0.12?
If yes, then we cannot use the new version; but instead, will need to update the examples to explicitly pass the `source_image_project` and `source_image_family` values.

Let me know - I'll make the changes accordingly.

**Comment by morgante** [comment]
@kunal-g Looks like we're making the change here: #148

**Comment by kunal-g** [comment]
> @kunal-g Looks like we're making the change here: #148
> 
> It's fine to drop support for Terraform 0.12 at this point.

OK, great. So, I'll wait for #148 to be merged. After that, let's rerun the tests.
Anything else I need to work on? Any other review comments you want me to address?

**Comment by morgante** [comment]
@kunal-g Everything else looks good.

**Review by morgante — APPROVED** [review]

**Comment by kunal-g** [comment]
@morgante now that #148 is merged, can you please merge this as well - if everything else looks good?

**Comment by morgante** [comment]
@kunal-g Can you rebase on master? This does look good.

**Comment by morgante** [comment]
@kunal-g Gentle ping, could you rebase?

**Review by naseemkullah — APPROVED** [review]

**Comment by kunal-g** [comment]
> @kunal-g Gentle ping, could you rebase?

Hi @morgante sorry, a bit tied up with work - will rebase + push over the weekend.

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate . 
Success!
 The configuration is valid.
terraform_validate ./examples/cloudrun 
Success!
 The configuration is valid.
terraform_validate ./examples/https-gke 
Success!
 The configuration is valid.
terraform_validate ./examples/https-gke/gke-node-port 
Error: 
Unsupported block type
  on main.tf line 121, in resource "kubernetes_replication_controller" "nginx":
 121:             
limits
 {
Blocks of type "limits" are not expected here. Did you mean to define argument
"limits"? If so, use the equals sign to assign it a value.
Error: 
Unsupported block type
  on main.tf line 126, in resource "kubernetes_replication_controller" "nginx":
 126:             
requests
 {
Blocks of type "requests" are not expected here. Did you mean to define
argument "requests"? If so, use the equals sign to assign it a value.
terraform_validate ./examples/https-redirect 
Success!
 The configuration is valid.
terraform_validate ./examples/mig-nat-http-lb 
Success!
 The configuration is valid.
terraform_validate ./examples/multi-backend-multi-mig-bucket-https-lb 
Success!
 The configuration is valid.
terraform_validate ./examples/multi-mig-http-lb 
Success!
 The configuration is valid.
terraform_validate ./examples/multiple-certs 
Success!
 The configuration is valid.
terraform_validate ./examples/shared-vpc 
Success!
 The configuration is valid.
terraform_validate ./modules/dynamic_backends 
Success!
 The configuration is valid.
terraform_validate ./modules/serverless_negs 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/cloudrun 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/https-redirect 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/mig_nat 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/multi_certs 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/multi_mig 
Success!
 The configuration is valid.
terraform_validate ./test/setup 
Success!
 The configuration is valid.
 ```

**Comment by kunal-g** [comment]
@morgante I have rebased to latest master and pushed - just as you asked.
There is no change in the code - but, because of the rebase, all the older commits are showing up as new.

Please have a look and let me know if anything needs to be fixed further.

**Review by morgante — APPROVED** [review]

**Comment by morgante** [comment]
@kunal-g Looks good, thanks for the patience + rebase.

**Comment by kunal-g** [comment]
> @kunal-g Looks good, thanks for the patience + rebase.

Glad I could help 😄

**Pull request merged**
